Kernel/Devices/Audio/IntelHDA/Controller.cpp

```diff
@@ -0,0 +1,342 @@
+    //         setting/clearing this bit such that the minimum link RESET# signal assertion
+    //         pulse width specification is met (see Section 5.5)."
+    microseconds_delay(100);
+    global_control |= GlobalControlFlag::ControllerReset;
+    m_controller_io_window->write32(ControllerRegister::GlobalControl, global_control);
+
+    // 3.3.7: "When the controller hardware is ready to begin operation, it will report a 1 in
+    //         this bit. Software must read a 1 from this bit before accessing any controller
+    //         registers."
+    TRY(wait_until(frame_delay_in_microseconds(1), controller_timeout_in_microseconds, [&]() {
+        global_control = m_controller_io_window->read32(ControllerRegister::GlobalControl);
+        return (global_control & GlobalControlFlag::ControllerReset) > 0;
+    }));
+
+    // 4.3 Codec Discovery:
+    // "The software must wait at least 521 us (25 frames) after reading CRST as a 1 before
+    // assuming that codecs have all made status change requests and have been registered
+    // by the controller."
+    microseconds_delay(frame_delay_in_microseconds(25));
+
+    dbgln_if(INTEL_HDA_DEBUG, "Controller reset");
+    return {};
+}
+
+LockRefPtr<AudioChannel> Controller::audio_channel(u32 index) const
+{
+    if (index != fixed_audio_channel_index)
+        return {};
+    return m_audio_channel;
+}
+
+ErrorOr<size_t> Controller::write(size_t channel_index, UserOrKernelBuffer const& data, size_t length)
+{
+    if (channel_index != fixed_audio_channel_index || !m_output_path)
+        return ENODEV;
+    return m_output_path->output_stream().write(data, length);
+}
+
+UNMAP_AFTER_INIT void Controller::detect_hardware_audio_channels(Badge<AudioManagement>)
+{
+    auto result = configure_output_route();
+    if (result.is_error()) {
+        dmesgln_pci(*this, "Failed to set up an output audio channel: {}", result.error());
+        return;
+    }
+
+    m_audio_channel = AudioChannel::must_create(*this, fixed_audio_channel_index);
+}
+
+ErrorOr<void> Controller::set_pcm_output_sample_rate(size_t channel_index, u32 samples_per_second_rate)
+{
+    if (channel_index != fixed_audio_channel_index || !m_output_path)
+        return ENODEV;
+
+    TRY(m_output_path->set_format({
+        .sample_rate = samples_per_second_rate,
+        .pcm_bits = OutputPath::fixed_pcm_bits,
+        .number_of_channels = OutputPath::fixed_channel_count,
+    }));
+    dmesgln_pci(*this, "Set output channel #{} PCM rate: {} Hz", channel_index, samples_per_second_rate);
+    return {};
+}
+
+ErrorOr<u32> Controller::get_pcm_output_sample_rate(size_t channel_index)
+{
+    if (channel_index != fixed_audio_channel_index || !m_output_path)
+        return ENODEV;
+
+    return m_output_path->output_stream().sample_rate();
+}
+
+ErrorOr<void> wait_until(size_t delay_in_microseconds, size_t timeout_in_microseconds, Function<ErrorOr<bool>()> condition)
```

Comment:
There are multiple instances of this pattern in the codebase. Ideally we would want to unify them sometime in the future. For now this looks OK.

Reply:
Yeah, `AHCIPort::wait_until_condition_met_or_timeout` comes to mind. There are multiple variants that work with either timeouts, number of retries, etc. but none of them are already generic. I opted to keep this one inside the HDA driver for now to not inflate the scope of my manual tests too much... agreed that we should unify this generic waiting logic in the future.